design-proposals/monitoring-code-refactor/monitoring-code-refactor.md

```diff
@@ -0,0 +1,223 @@
+# Monitoring code refactor
+
+**Authors:** João Vilaça <jvilaca@redhat.com>
+
+## What
+
+This design document is proposing a code refactor for the monitoring logic in
+all KubeVirt components. The goal is to have a centralized place for the
+creation of monitoring resources, a consistent monitoring package between all
+projects, a code struct that is easy to maintain and evolve, while moving closer
+to the Kubernetes metric implementation style.
+
+## Why
+
+It’s important to refactor the monitoring code to separate monitoring logic from
+business logic. It will make the codebase more modular, readable, and
+maintainable. It will reduce complexity, reduce the risk of introducing errors,
+and allow us to test each component independently.
+
+Right now, every change regarding monitoring needs to happen at many different
+places. For example, when adding the linter for metrics names, it was necessary
+to go to all KubeVirt components repositories and implement the linter
+validation in each. This is demanding, very time-consuming, requires going
+through multiple duplicated code reviews, uses many CI resources, and makes it
+mentally harder to decide to change anything.
```

Comment:
I disagree with the statement that it "makes it mentally harder to decide to change anything."
We never thought about this in this way.

I still think that there are significant advantages for having the monitoring tied to the specific operator that its related to.

We have a hard time in the past with other operators that we struggled to understand where their monitoring come from and how they are calculated, when they were not in the same repository.

Reply:
> I disagree with the statement that it "makes it mentally harder to decide to change anything."

Don't you agree that adding code to multiple repositories, having to code-review and get feedback from multiple of them, while having to align the logic between them brings complexities and intimidates from introducing more complex monitoring logic?

I think what @machadovilaca wrote there makes a lot of sense, and my past experience is similar.

> I still think that there are significant advantages for having the monitoring tied to the specific operator that its 
related to.

Can you outline the advantages?

> We have a hard time in the past with other operators that we struggled to understand where their monitoring come from and how they are calculated, when they were not in the same repository.

I think entirely the opposite.
When the monitoring code is part of the core operator in many different repositories it's very hard to reason about it. When the code resides at a centralized repository it's easy to find and understand. We can also make sure it's well-documented.

Reply:
Could you please provide your thought on the following scenario:
- Someone is trying to add monitoring code to repos A, B and C.
- In repo A, the community accepts the changes.
- In repo B, the community accepts the changes in general, but demands several core changes.
- In repo C, there's a significant push-back regarding the code and they don't want to adopt it without significant changes.

Reply:
Please try to think of it differently.
When you take out the monitoring from the repo, then you also take away the maintainers and the repo community control over it.
Can you really prevent each repo from adding new metrics and alerts that are asked for? You may end up with having monitoring for the same repo in several places.
Yes, Adding functionality that we need to several repos has its challenges, but if its good, like the linter, the docs generator, tests and so on, the community will accept it. 
Another point is that getting the feedback and requests from the different operators makes the code much better at the end.